File: Packages/Yage/Sources/Yage/Capabilities/LinearMovement.swift
import SwiftUI

public class LinearMovement: Capability {
    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard let subject, subject.state.canMove else { return }
        let distance = movement(after: time)
        let newPosition = subject.frame.origin.offset(by: distance)
        subject.frame.origin = newPosition
    }

    func movement(after time: TimeInterval) -> CGPoint {
        guard let subject else { return .zero }
        return CGPoint(
            x: subject.direction.dx * subject.speed * time,
            y: subject.direction.dy * subject.speed * time
        )
    }
}

public extension Entity {
    var movement: LinearMovement? {
        capability(for: LinearMovement.self)
    }
}

private extension EntityState {
    var canMove: Bool {
        self == .move || self == .freeFall
    }
}

File: Sources/windows/yage/capabilities/linear_movement.py
from typing import List
from yage.models.capability import Capability
from yage.models.collisions import Collision
from yage.models.entity_state import EntityState
from yage.utils.geometry import Point


class LinearMovement(Capability):
    def __init__(self, subject):
        super().__init__(subject)
        subject.movement = self

    def do_update(self, collisions: List[Collision], time: float):
        if not self.can_move:
            return
        distance = self.movement_given_time(time)
        new_position = self.subject.frame.origin.offset(point=distance)
        self.subject.frame.origin = new_position

    def movement_given_time(self, time: float) -> Point:
        return Point(
            self.subject.direction.dx * self.subject.speed * time,
            self.subject.direction.dy * self.subject.speed * time
        )

    @property
    def can_move(self) -> bool:
        return self.subject.state in [EntityState.MOVE, EntityState.FREE_FALL]
